lint_lib/lint.py

```diff
@@ -0,0 +1,272 @@
+        else:
+            found_input = set(item_dict.items())
+
+        first_line = seen_items.setdefault(frozenset(found_input), item.lineno)
+        if first_line is not None and first_line != item.lineno:
+            print(
+                f"Duplicate item in {path} at line {item.lineno} previously seen on line {first_line}"
+            )
+
+    return [dict(x.data) for x in parsed]
+
+
+def lint_encoding_test(path: str) -> None:
+    parsed = lint_dat_format(
+        path,
+        None,
+        b"data",
+        expected_headers=[b"data", b"encoding"],
+        input_headers={b"data"},
+    )
+    if not parsed:
+        return
+
+
+def lint_encoding_tests(path: str) -> None:
+    for root, dirs, files in os.walk(path):
+        for file in sorted(files):
+            if not file.endswith(".dat"):
+                continue
+            lint_encoding_test(clean_path(join(root, file)))
+
+
+def lint_tokenizer_test(path: str) -> None:
+    all_keys = {
+        "description",
+        "input",
+        "output",
+        "initialStates",
+        "lastStartTag",
+        "ignoreErrorOrder",
+        "doubleEscaped",
+        "errors",
+    }
+    required = {"input", "output"}
+    with codecs.open(path, "r", "utf-8") as fp:
+        parsed = json.load(fp)
+    if not parsed:
+        return
+    if not isinstance(parsed, dict):
+        print("Top-level must be an object in %s" % path)
+        return
+    for test_group in parsed.values():
+        if not isinstance(test_group, list):
+            print("Test groups must be a lists in %s" % path)
+            continue
+        for test in test_group:
+            if "doubleEscaped" in test and test["doubleEscaped"] is True:
+                test = unescape_json(test)
+            keys = set(test.keys())
+            if not (required <= keys):
+                print(
+                    "missing test properties {!r} in {}".format(required - keys, path)
+                )
+            if not (keys <= all_keys):
+                print(
+                    "unknown test properties {!r} in {}".format(keys - all_keys, path)
+                )
+
+
+def lint_tokenizer_tests(path: str) -> None:
+    for root, dirs, files in os.walk(path):
+        for file in sorted(files):
+            if not file.endswith(".test"):
+                continue
+            lint_tokenizer_test(clean_path(join(root, file)))
+
+
+def lint_tree_construction_test(path: str) -> None:
+    parsed = lint_dat_format(
+        path,
+        "utf-8",
+        "data",
+        expected_headers=[
+            "data",
+            "errors",
+            "new-errors",
+            "document-fragment",
+            "script-off",
+            "script-on",
+            "document",
+        ],
+        input_headers={
+            "data",
+            "document-fragment",
+            "script-on",
+            "script-off",
+        },
+    )
+    if not parsed:
+        return
```

Comment:
If this does something, it should do it in a less cryptic way

Reply:
@Ms2ger PTAL at the comments that are there now